Source/DafnyCore/Verifier/ProofDependency.cs

```diff
@@ -162,24 +168,40 @@ public class CallDependency : ProofDependency {
   }
 }
 
+public enum AssumptionForm {
```

Comment:
This enum barely seems used. The only check on its value I saw was

```
if (assumeDep.Form != AssumptionForm.AssumeStatement && assumeDep.Form != AssumptionForm.AssertStatement)
```

In which case I think you could have used a bool instead of an enum. I imagine you have future intentions for this enum but I suggest only writing code that is useful for the current set of tests, and adding more later.

Also, the enum seems to mirror fields of AST node types, so consider using references to those instead of an enum, for example like this:

```c#
interface IHasAssumptions : INode { 
  public X AnswerQuestionAboutAssumptionField(string fieldName);
}

class ComprehensionExpr : IHasAssumptions { 
  public X AnswerQuestionAboutAssumptionField(string fieldName) {
    if (fieldName == nameof(Range)) { 
      ..
    }
  }
}

record Assumption(IHasAssumptions Container, string FieldName)
```

Reply:
It's true, that enum is defined as it is in preparation for broader use in the future. I don't think a boolean would be appropriate here, because there are a number of different decisions I expect we may make based on the type of assumption in the future, and that would lead to the need for substantial refactoring in the future.

I would leave out the unused alternatives in the assumption for now except that it needs to be filled in at every creation site. Given that there is an `Other` alternative, I could leave out all of the other alternatives that are (currently) never checked and use `Other` for all of them.

Reply:
It's probably also worth noting that, although these are very similar to AST nodes, they don't map directly. For example, `ForallEnsures` and `ForallRange` describe aspects of a `forall` statement that are, themselves, only `Expression` values. If we were to use the `ForallStmt` node, we wouldn't know which part of it this assumption referred to. If we were to use the ensures or range expressions, we wouldn't know where those expressions came from.

Reply:
I'm fine with an enum, it's like a named boolean. Thanks for explaining that.

Reply:
> It's probably also worth noting that, although these are very similar to AST nodes, they don't map directly. For example, ForallEnsures and ForallRange describe aspects of a forall statement that are, themselves, only Expression values. If we were to use the ForallStmt node, we wouldn't know which part of it this assumption referred to. If we were to use the ensures or range expressions, we wouldn't know where those expressions came from.

I agree, that's why I said the enum values seems to mirror _fields_ of AST node types, and why the record I suggested 
```
record Assumption(IHasAssumptions Container, string FieldName)
```
includes a FieldName property.


> I don't think a boolean would be appropriate here, because there are a number of different decisions I expect we may make based on the type of assumption in the future, and that would lead to the need for substantial refactoring in the future.

Given the "I expect", are you saying you do think a boolean would be appropriate right now?

What do you mean with substantial refactoring? Changing a boolean into something else? What's substantial about that? Writing code is cheap. However, how can I review code that is meant to do things not specified or tested in the current PR? And how will you ask a future reviewer to review the whole set of changes that were accumulated of several PRs, once it has finally reached a state where each piece is fully used?

Reply:
The "substantial refactoring" I mentioned has to do with putting behavior in the AST nodes now and moving it back out later, not just changing the type of a field. But I think that may have been a misunderstanding of what you were suggesting. For the moment, I'd be happy to move to `WarnUnused` and `IgnoreUnused`, or something like that.

Reply:
> For the moment, I'd be happy to move to WarnUnused and IgnoreUnused, or something like that.

Thanks !